src/main/scala/com/google/cloud/spark/bigquery/BigQueryWriteHelper.scala

```diff
@@ -0,0 +1,147 @@
+      gcsPath = new Path(gcsPathOption.get)
+      val fs = gcsPath.getFileSystem(conf)
+      needNewPath = fs.exists(gcsPath) // if teh path exists for some reason, then retry
+    }
+
+    gcsPath
+  }
+
+  val saveModeToWriteDisposition = Map(
+    SaveMode.Append -> JobInfo.WriteDisposition.WRITE_APPEND,
+    SaveMode.Overwrite -> JobInfo.WriteDisposition.WRITE_TRUNCATE)
+
+  def writeDataFrameToBigQuery: Unit = {
+    try {
+      // based on pmkc's suggestion at https://git.io/JeWRt
+      Runtime.getRuntime.addShutdownHook(createTemporaryPathDeleter)
+
+      val format = options.intermediateFormat.getType.toLowerCase
+      data.write.format(format).save(temporaryGcsPath.toString)
+      loadDataToBigQuery
+    } catch {
+      case e: Exception => throw new RuntimeException("Failed to write to BigQuery", e)
+    } finally {
+      cleanTempraryGcsPath
+    }
+  }
+
+  def loadDataToBigQuery(): Unit = {
+    val fs = temporaryGcsPath.getFileSystem(conf)
+    val sourceUris = ToIterator(fs.listFiles(temporaryGcsPath, false))
+      .map(_.getPath.toString)
+      .filter(!_.endsWith("_SUCCESS"))
+      .toList
+      .asJava
+
+    val jobConfiguration = LoadJobConfiguration.newBuilder(
+      options.tableId, sourceUris, options.intermediateFormat)
+      .setCreateDisposition(JobInfo.CreateDisposition.CREATE_IF_NEEDED)
+      .setWriteDisposition(saveModeToWriteDisposition(saveMode))
+      .build
+
+    val jobInfo = JobInfo.of(jobConfiguration)
+    val job = bigQuery.create(jobInfo)
+
+    logger.info(s"Submitted load to ${options.tableId}. jobId: ${job.getJobId}")
+    // TODO(davidrab): add retry options
+    val finishedJob = job.waitFor()
+    if (finishedJob.getStatus.getError != null) {
+      throw new BigQueryException(
+        BaseHttpServiceException.UNKNOWN_CODE,
+        s"""Failed to load to ${friendlyTableName} in job ${job.getJobId}. BigQuery error was
+           |${finishedJob.getStatus.getError.getMessage}""".stripMargin.replace('\n', ' '),
+        finishedJob.getStatus.getError)
+    } else {
+      logger.info(s"Done loading to ${friendlyTableName}. jobId: ${job.getJobId}")
+    }
+  }
+
+  def friendlyTableName: String = BigQueryUtil.friendlyTableName(options.tableId)
+
+  def cleanTempraryGcsPath: Unit = {
+    // TODO(davidrab): add flag to disable the deletion?
+    createTemporaryPathDeleter.deletePath
+  }
+
+  private def createTemporaryPathDeleter = HdfsPathDeleter(temporaryGcsPath, conf)
+
+  def verifySaveMode: Unit = {
+    if (saveMode == SaveMode.ErrorIfExists || saveMode == SaveMode.Ignore) {
+      throw new UnsupportedOperationException(s"SaveMode $saveMode is not supported")
+    }
+  }
+}
+
+/**
+ * Resposible for recursively deleting an HDFS path. Implementing Thread in order
+ * to act as shutdown hook
+ * @param path the path to delete
+ * @param conf the hadoop configuration
+ */
+case class HdfsPathDeleter(path: Path, conf: Configuration) extends Thread with StrictLogging {
```

Comment:
nit: never call GCS HDFS.
It is either HCFS (compatible) or just filesystem

In this case I might say IntermediateDataCleaner or something

Reply:
Fixed in https://github.com/GoogleCloudPlatform/spark-bigquery-connector/pull/66/commits/d34dc2094abdee77ad48558ebe9490e0ff901896